File: backend/routers/interfaces/l2tpv3.py
"""
L2TPv3 Interface Configuration Endpoints

All L2TPv3 (Layer 2 Tunnel Protocol Version 3) interface endpoints for VyOS configuration.
"""

import inspect
import logging
from typing import Dict, List, Optional, Any

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field, ConfigDict
from starlette.concurrency import run_in_threadpool

from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
from session_vyos_service import get_session_vyos_service

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/l2tpv3", tags=["l2tpv3-interface"])


# Stub functions for backwards compatibility with app.py
def set_device_registry(registry):
    """Legacy function - no longer used."""
    pass


def set_configured_device_name(name):
    """Legacy function - no longer used."""
    pass


# ============================================================================
# Request / Response Models
# ============================================================================


class BatchOperation(BaseModel):
    op: str = Field(..., description="Operation name")
    value: Optional[str] = Field(None, description="Operation value (if required)")


class BatchRequest(BaseModel):
    interface: str = Field(..., description="Interface name (e.g., l2tpeth0)")
    operations: List[BatchOperation]


class VyOSResponse(BaseModel):
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


class L2TPv3InterfaceConfig(BaseModel):
    name: str
    type: str
    addresses: List[str] = Field(default_factory=list)
    description: Optional[str] = None
    vrf: Optional[str] = None
    mtu: Optional[str] = None
    disable: Optional[bool] = None
    # L2TPv3-specific tunnel settings
    remote: Optional[str] = None
    source_address: Optional[str] = None
    tunnel_id: Optional[str] = None
    peer_tunnel_id: Optional[str] = None
    session_id: Optional[str] = None
    peer_session_id: Optional[str] = None
    encapsulation: Optional[str] = None
    destination_port: Optional[str] = None
    source_port: Optional[str] = None
    # IP settings
    ip_adjust_mss: Optional[str] = None
    ip_arp_cache_timeout: Optional[str] = None
    ip_disable_arp_filter: Optional[bool] = None
    ip_disable_forwarding: Optional[bool] = None
    ip_enable_arp_accept: Optional[bool] = None
    ip_enable_arp_announce: Optional[bool] = None
    ip_enable_arp_ignore: Optional[bool] = None
    ip_enable_directed_broadcast: Optional[bool] = None
    ip_enable_proxy_arp: Optional[bool] = None
    ip_proxy_arp_pvlan: Optional[bool] = None
    ip_source_validation: Optional[str] = None
    # IPv6 settings
    ipv6_accept_dad: Optional[str] = None
    ipv6_address_autoconf: Optional[bool] = None
    ipv6_address_eui64: List[str] = Field(default_factory=list)
    ipv6_address_interface_identifier: Optional[str] = None
    ipv6_address_no_default_link_local: Optional[bool] = None
    ipv6_adjust_mss: Optional[str] = None
    ipv6_base_reachable_time: Optional[str] = None
    ipv6_disable_forwarding: Optional[bool] = None
    ipv6_dup_addr_detect_transmits: Optional[str] = None
    ipv6_source_validation: Optional[str] = None
    # Mirror
    mirror_ingress: Optional[str] = None
    mirror_egress: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)


class L2TPv3InterfacesConfigResponse(BaseModel):
    interfaces: List[L2TPv3InterfaceConfig] = Field(default_factory=list)
    total: int = 0
    by_type: Dict[str, int] = Field(default_factory=dict)
    by_vrf: Dict[str, int] = Field(default_factory=dict)


# ============================================================================
# Endpoints
# ============================================================================


@router.get("/capabilities")
async def get_capabilities(request: Request) -> Dict[str, Any]:
    """Return version-aware feature capabilities for L2TPv3 interfaces."""
    await require_read_permission(request, FeatureGroup.INTERFACES)
    service = get_session_vyos_service(request)
    from vyos_builders.interfaces.l2tpv3 import L2TPv3InterfaceBuilderMixin
    builder = L2TPv3InterfaceBuilderMixin(version=service.get_version())
    return builder.get_capabilities()


@router.get("/config", response_model=L2TPv3InterfacesConfigResponse)
async def get_config(http_request: Request, refresh: bool = False) -> L2TPv3InterfacesConfigResponse:
    """Get all L2TPv3 interface configurations from VyOS."""
    await require_read_permission(http_request, FeatureGroup.INTERFACES)
    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh)
        raw_config = full_config.get("interfaces", {}).get("l2tpv3", {})

        from vyos_mappers.interfaces.l2tpv3_versions import get_l2tpv3_mapper
        mapper = get_l2tpv3_mapper(service.get_version())
        parsed = mapper.parse_interfaces_of_type(raw_config)
        return L2TPv3InterfacesConfigResponse(**parsed)
    except Exception:
        logger.exception("Unhandled error in get_config")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/batch", response_model=VyOSResponse)
async def batch_configure(http_request: Request, request: BatchRequest) -> VyOSResponse:
    """
    Configure an L2TPv3 interface using batch operations.

    **Supported operations:**
    | Operation | Value | Description |
    |-----------|-------|-------------|
    | `set_interface_description` | Yes | Set description |
    | `delete_interface_description` | No | Remove description |
    | `set_interface_address` | Yes | Add IP address (CIDR) |
    | `delete_interface_address` | Yes | Remove IP address |
    | `set_interface_mtu` | Yes | Set MTU (68-16000, default 1488) |
    | `delete_interface_mtu` | No | Reset MTU to default |
    | `set_interface_disable` | No | Administratively disable |
    | `delete_interface_disable` | No | Re-enable interface |
    | `set_interface_vrf` | Yes | Assign to VRF |
    | `delete_interface_vrf` | No | Remove VRF assignment |
    | `set_remote` | Yes | Set tunnel remote address |
    | `delete_remote` | No | Remove tunnel remote address |
    | `set_source_address` | Yes | Set source IP address |
    | `delete_source_address` | No | Remove source address |
    | `set_tunnel_id` | Yes | Set local tunnel ID (1-429496729) |
    | `delete_tunnel_id` | No | Remove tunnel ID |
    | `set_peer_tunnel_id` | Yes | Set peer tunnel ID (1-429496729) |
    | `delete_peer_tunnel_id` | No | Remove peer tunnel ID |
    | `set_session_id` | Yes | Set session ID (1-429496729) |
    | `delete_session_id` | No | Remove session ID |
    | `set_peer_session_id` | Yes | Set peer session ID (1-429496729) |
    | `delete_peer_session_id` | No | Remove peer session ID |
    | `set_encapsulation` | Yes | Set encapsulation (udp/ip) |
    | `delete_encapsulation` | No | Remove encapsulation |
    | `set_destination_port` | Yes | Set UDP destination port (1-65535) |
    | `delete_destination_port` | No | Remove destination port |
    | `set_source_port` | Yes | Set UDP source port (1-65535) |
    | `delete_source_port` | No | Remove source port |
    | `set_ip_adjust_mss` | Yes | Set TCP MSS (clamp-mss-to-pmtu or 536-65535) |
    | `delete_ip_adjust_mss` | No | Remove TCP MSS |
    | `set_ip_arp_cache_timeout` | Yes | Set ARP cache timeout (1-86400) |
    | `delete_ip_arp_cache_timeout` | No | Remove ARP cache timeout |
    | `set_ip_disable_arp_filter` | No | Disable ARP filter |
    | `delete_ip_disable_arp_filter` | No | Enable ARP filter |
    | `set_ip_disable_forwarding` | No | Disable IPv4 forwarding |
    | `delete_ip_disable_forwarding` | No | Enable IPv4 forwarding |
    | `set_ip_enable_arp_accept` | No | Enable ARP accept |
    | `delete_ip_enable_arp_accept` | No | Disable ARP accept |
    | `set_ip_enable_arp_announce` | No | Enable ARP announce |
    | `delete_ip_enable_arp_announce` | No | Disable ARP announce |
    | `set_ip_enable_arp_ignore` | No | Enable ARP ignore |
    | `delete_ip_enable_arp_ignore` | No | Disable ARP ignore |
    | `set_ip_enable_directed_broadcast` | No | Enable directed broadcast |
    | `delete_ip_enable_directed_broadcast` | No | Disable directed broadcast |
    | `set_ip_enable_proxy_arp` | No | Enable proxy ARP |
    | `delete_ip_enable_proxy_arp` | No | Disable proxy ARP |
    | `set_ip_proxy_arp_pvlan` | No | Enable private VLAN proxy ARP |
    | `delete_ip_proxy_arp_pvlan` | No | Disable private VLAN proxy ARP |
    | `set_ip_source_validation` | Yes | Source validation (strict/loose/disable) |
    | `delete_ip_source_validation` | No | Remove source validation |
    | `set_ipv6_accept_dad` | Yes | Set DAD mode (0/1/2) |
    | `delete_ipv6_accept_dad` | No | Remove DAD setting |
    | `set_ipv6_address_autoconf` | No | Enable SLAAC autoconfig |
    | `delete_ipv6_address_autoconf` | No | Disable SLAAC autoconfig |
    | `set_ipv6_address_eui64` | Yes | Add EUI-64 prefix |
    | `delete_ipv6_address_eui64` | Yes | Remove EUI-64 prefix |
    | `set_ipv6_address_interface_identifier` | Yes | Set SLAAC interface identifier |
    | `delete_ipv6_address_interface_identifier` | No | Remove interface identifier |
    | `set_ipv6_address_no_default_link_local` | No | Remove default link-local |
    | `delete_ipv6_address_no_default_link_local` | No | Restore default link-local |
    | `set_ipv6_adjust_mss` | Yes | Set IPv6 TCP MSS |
    | `delete_ipv6_adjust_mss` | No | Remove IPv6 TCP MSS |
    | `set_ipv6_base_reachable_time` | Yes | Set base reachable time (1-86400) |
    | `delete_ipv6_base_reachable_time` | No | Remove base reachable time |
    | `set_ipv6_disable_forwarding` | No | Disable IPv6 forwarding |
    | `delete_ipv6_disable_forwarding` | No | Enable IPv6 forwarding |
    | `set_ipv6_dup_addr_detect_transmits` | Yes | Set DAD NS count |
    | `delete_ipv6_dup_addr_detect_transmits` | No | Remove DAD NS count |
    | `set_ipv6_source_validation` | Yes | IPv6 source validation (strict/loose/disable) |
    | `delete_ipv6_source_validation` | No | Remove IPv6 source validation |
    | `set_mirror_ingress` | Yes | Mirror ingress to interface |
    | `delete_mirror_ingress` | No | Remove ingress mirror |
    | `set_mirror_egress` | Yes | Mirror egress to interface |
    | `delete_mirror_egress` | No | Remove egress mirror |
    | `delete_interface` | No | Delete entire interface |
    """
    await require_write_permission(http_request, FeatureGroup.INTERFACES)

    try:
        service = get_session_vyos_service(http_request)
        batch = service.create_l2tpv3_batch()

        for op in request.operations:
            if op.op in batch._INTERNAL_BUILDER_METHODS:
                raise HTTPException(
                    status_code=400,
                    detail=f"Operation '{op.op}' is not a valid interface operation",
                )

            method = getattr(batch, op.op, None)
            if method is None:
                raise HTTPException(
                    status_code=400,
                    detail=f"Unsupported operation: {op.op}",
                )

            sig = inspect.signature(method)
            params = [p for p in sig.parameters.keys() if p != "self"]

            if len(params) == 1:
                method(request.interface)
            elif len(params) == 2:
                if op.value is None:
                    raise HTTPException(
                        status_code=400,
                        detail=f"Operation '{op.op}' requires a value",
                    )
                method(request.interface, op.value)
            else:
                raise HTTPException(
                    status_code=400,
                    detail=f"Operation '{op.op}' has unexpected signature",
                )

        response = service.execute_batch(batch)
        return VyOSResponse(
            success=response.status == 200,
            data=response.result if isinstance(response.result, dict) else None,
            error=response.error if response.error else None,
        )
    except HTTPException:
        raise
    except Exception:
        logger.exception("Unhandled error in batch_configure")
        raise HTTPException(status_code=500, detail="Internal server error")

File: frontend/src/lib/api/l2tpv3.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface L2TPv3Interface {
  name: string;
  type: string;
  addresses: string[];
  description: string | null;
  vrf: string | null;
  mtu: string | null;
  disable: boolean | null;
  // L2TPv3-specific tunnel settings
  remote: string | null;
  source_address: string | null;
  tunnel_id: string | null;
  peer_tunnel_id: string | null;
  session_id: string | null;
  peer_session_id: string | null;
  encapsulation: string | null;
  destination_port: string | null;
  source_port: string | null;
  // IP settings
  ip_adjust_mss: string | null;
  ip_arp_cache_timeout: string | null;
  ip_disable_arp_filter: boolean | null;
  ip_disable_forwarding: boolean | null;
  ip_enable_arp_accept: boolean | null;
  ip_enable_arp_announce: boolean | null;
  ip_enable_arp_ignore: boolean | null;
  ip_enable_directed_broadcast: boolean | null;
  ip_enable_proxy_arp: boolean | null;
  ip_proxy_arp_pvlan: boolean | null;
  ip_source_validation: string | null;
  // IPv6 settings
  ipv6_accept_dad: string | null;
  ipv6_address_autoconf: boolean | null;
  ipv6_address_eui64: string[];
  ipv6_address_interface_identifier: string | null;
  ipv6_address_no_default_link_local: boolean | null;
  ipv6_adjust_mss: string | null;
  ipv6_base_reachable_time: string | null;
  ipv6_disable_forwarding: boolean | null;
  ipv6_dup_addr_detect_transmits: string | null;
  ipv6_source_validation: string | null;
  // Mirror
  mirror_ingress: string | null;
  mirror_egress: string | null;
}

export interface L2TPv3ConfigResponse {
  interfaces: L2TPv3Interface[];
  total: number;
  by_type: Record<string, number>;
  by_vrf: Record<string, number>;
}

export interface L2TPv3Capabilities {
  version: string;
  features: Record<string, { supported: boolean; description: string }>;
  version_info?: {
    is_1_4: boolean;
    is_1_5: boolean;
  };
  instance_name?: string;
  instance_id?: string;
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

export interface L2TPv3BatchOperation {
  op: string;
  value?: string;
}

// ============================================================================
// API Service
// ============================================================================

class L2TPv3Service {
  async getCapabilities(): Promise<L2TPv3Capabilities> {
    return apiClient.get<L2TPv3Capabilities>("/vyos/l2tpv3/capabilities");
  }

  async getConfig(refresh: boolean = false): Promise<L2TPv3ConfigResponse> {
    return apiClient.get<L2TPv3ConfigResponse>("/vyos/l2tpv3/config", {
      refresh: refresh.toString(),
    });
  }

  async refreshConfig(): Promise<VyOSResponse> {
    return apiClient.post("/vyos/config/refresh");
  }

  async batchConfigure(
    interfaceName: string,
    operations: L2TPv3BatchOperation[]
  ): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/l2tpv3/batch", {
      interface: interfaceName,
      operations,
    });
    await this.refreshConfig();
    return result;
  }

  async createInterface(config: {
    name: string;
    description?: string;
    addresses?: string[];
    mtu?: string;
    vrf?: string;
    disabled?: boolean;
    // L2TPv3-specific
    remote?: string;
    source_address?: string;
    tunnel_id?: string;
    peer_tunnel_id?: string;
    session_id?: string;
    peer_session_id?: string;
    encapsulation?: string;
    destination_port?: string;
    source_port?: string;
    // IP settings
    ip_adjust_mss?: string;
    ip_arp_cache_timeout?: string;
    ip_disable_arp_filter?: boolean;
    ip_disable_forwarding?: boolean;
    ip_enable_arp_accept?: boolean;
    ip_enable_arp_announce?: boolean;
    ip_enable_arp_ignore?: boolean;
    ip_enable_directed_broadcast?: boolean;
    ip_enable_proxy_arp?: boolean;
    ip_proxy_arp_pvlan?: boolean;
    ip_source_validation?: string;
    // IPv6 settings
    ipv6_accept_dad?: string;
    ipv6_adjust_mss?: string;
    ipv6_base_reachable_time?: string;
    ipv6_disable_forwarding?: boolean;
    ipv6_dup_addr_detect_transmits?: string;
    ipv6_source_validation?: string;
    ipv6_address_autoconf?: boolean;
    ipv6_address_eui64?: string[];
    ipv6_address_no_default_link_local?: boolean;
    ipv6_address_interface_identifier?: string;
    // Mirror
    mirror_ingress?: string;
    mirror_egress?: string;
  }): Promise<VyOSResponse> {
    const operations: L2TPv3BatchOperation[] = [];

    // L2TPv3-specific tunnel settings
    if (config.remote) operations.push({ op: "set_remote", value: config.remote });
    if (config.source_address) operations.push({ op: "set_source_address", value: config.source_address });
    if (config.tunnel_id) operations.push({ op: "set_tunnel_id", value: config.tunnel_id });
    if (config.peer_tunnel_id) operations.push({ op: "set_peer_tunnel_id", value: config.peer_tunnel_id });
    if (config.session_id) operations.push({ op: "set_session_id", value: config.session_id });
    if (config.peer_session_id) operations.push({ op: "set_peer_session_id", value: config.peer_session_id });
    if (config.encapsulation) operations.push({ op: "set_encapsulation", value: config.encapsulation });
    if (config.destination_port) operations.push({ op: "set_destination_port", value: config.destination_port });
    if (config.source_port) operations.push({ op: "set_source_port", value: config.source_port });

    // Basic
    if (config.description) operations.push({ op: "set_interface_description", value: config.description });
    if (config.addresses) {
      for (const addr of config.addresses) {
        operations.push({ op: "set_interface_address", value: addr });
      }
    }
    if (config.mtu) operations.push({ op: "set_interface_mtu", value: config.mtu });
    if (config.vrf) operations.push({ op: "set_interface_vrf", value: config.vrf });
    if (config.disabled) operations.push({ op: "set_interface_disable" });

    // IP settings
    if (config.ip_adjust_mss) operations.push({ op: "set_ip_adjust_mss", value: config.ip_adjust_mss });
    if (config.ip_arp_cache_timeout) operations.push({ op: "set_ip_arp_cache_timeout", value: config.ip_arp_cache_timeout });
    if (config.ip_disable_arp_filter) operations.push({ op: "set_ip_disable_arp_filter" });
    if (config.ip_disable_forwarding) operations.push({ op: "set_ip_disable_forwarding" });
    if (config.ip_enable_arp_accept) operations.push({ op: "set_ip_enable_arp_accept" });
    if (config.ip_enable_arp_announce) operations.push({ op: "set_ip_enable_arp_announce" });
    if (config.ip_enable_arp_ignore) operations.push({ op: "set_ip_enable_arp_ignore" });
    if (config.ip_enable_directed_broadcast) operations.push({ op: "set_ip_enable_directed_broadcast" });
    if (config.ip_enable_proxy_arp) operations.push({ op: "set_ip_enable_proxy_arp" });
    if (config.ip_proxy_arp_pvlan) operations.push({ op: "set_ip_proxy_arp_pvlan" });
    if (config.ip_source_validation) operations.push({ op: "set_ip_source_validation", value: config.ip_source_validation });

    // IPv6 settings
    if (config.ipv6_accept_dad) operations.push({ op: "set_ipv6_accept_dad", value: config.ipv6_accept_dad });
    if (config.ipv6_adjust_mss) operations.push({ op: "set_ipv6_adjust_mss", value: config.ipv6_adjust_mss });
    if (config.ipv6_base_reachable_time) operations.push({ op: "set_ipv6_base_reachable_time", value: config.ipv6_base_reachable_time });
    if (config.ipv6_disable_forwarding) operations.push({ op: "set_ipv6_disable_forwarding" });
    if (config.ipv6_dup_addr_detect_transmits) operations.push({ op: "set_ipv6_dup_addr_detect_transmits", value: config.ipv6_dup_addr_detect_transmits });
    if (config.ipv6_source_validation) operations.push({ op: "set_ipv6_source_validation", value: config.ipv6_source_validation });
    if (config.ipv6_address_autoconf) operations.push({ op: "set_ipv6_address_autoconf" });
    if (config.ipv6_address_eui64) {
      for (const prefix of config.ipv6_address_eui64) {
        operations.push({ op: "set_ipv6_address_eui64", value: prefix });
      }
    }
    if (config.ipv6_address_no_default_link_local) operations.push({ op: "set_ipv6_address_no_default_link_local" });
    if (config.ipv6_address_interface_identifier) operations.push({ op: "set_ipv6_address_interface_identifier", value: config.ipv6_address_interface_identifier });

    // Mirror
    if (config.mirror_ingress) operations.push({ op: "set_mirror_ingress", value: config.mirror_ingress });
    if (config.mirror_egress) operations.push({ op: "set_mirror_egress", value: config.mirror_egress });

    return this.batchConfigure(config.name, operations);
  }

  async updateInterface(
    name: string,
    current: L2TPv3Interface,
    updated: {
      description?: string | null;
      addresses?: string[];
      mtu?: string | null;
      vrf?: string | null;
      disabled?: boolean | null;
      remote?: string | null;
      source_address?: string | null;
      tunnel_id?: string | null;
      peer_tunnel_id?: string | null;
      session_id?: string | null;
      peer_session_id?: string | null;
      encapsulation?: string | null;
      destination_port?: string | null;
      source_port?: string | null;
      ip_adjust_mss?: string | null;
      ip_arp_cache_timeout?: string | null;
      ip_disable_arp_filter?: boolean | null;
      ip_disable_forwarding?: boolean | null;
      ip_enable_arp_accept?: boolean | null;
      ip_enable_arp_announce?: boolean | null;
      ip_enable_arp_ignore?: boolean | null;
      ip_enable_directed_broadcast?: boolean | null;
      ip_enable_proxy_arp?: boolean | null;
      ip_proxy_arp_pvlan?: boolean | null;
      ip_source_validation?: string | null;
      ipv6_accept_dad?: string | null;
      ipv6_adjust_mss?: string | null;
      ipv6_base_reachable_time?: string | null;
      ipv6_disable_forwarding?: boolean | null;
      ipv6_dup_addr_detect_transmits?: string | null;
      ipv6_source_validation?: string | null;
      ipv6_address_autoconf?: boolean | null;
      ipv6_address_eui64?: string[];
      ipv6_address_no_default_link_local?: boolean | null;
      ipv6_address_interface_identifier?: string | null;
      mirror_ingress?: string | null;
      mirror_egress?: string | null;
    }
  ): Promise<VyOSResponse> {
    const operations: L2TPv3BatchOperation[] = [];

    // Simple string fields
    const stringFields: { key: keyof typeof updated; setOp: string; deleteOp: string; currentVal: string | null }[] = [
      { key: "description", setOp: "set_interface_description", deleteOp: "delete_interface_description", currentVal: current.description },
      { key: "mtu", setOp: "set_interface_mtu", deleteOp: "delete_interface_mtu", currentVal: current.mtu },
      { key: "vrf", setOp: "set_interface_vrf", deleteOp: "delete_interface_vrf", currentVal: current.vrf },
      { key: "remote", setOp: "set_remote", deleteOp: "delete_remote", currentVal: current.remote },
      { key: "source_address", setOp: "set_source_address", deleteOp: "delete_source_address", currentVal: current.source_address },
      { key: "tunnel_id", setOp: "set_tunnel_id", deleteOp: "delete_tunnel_id", currentVal: current.tunnel_id },
      { key: "peer_tunnel_id", setOp: "set_peer_tunnel_id", deleteOp: "delete_peer_tunnel_id", currentVal: current.peer_tunnel_id },
      { key: "session_id", setOp: "set_session_id", deleteOp: "delete_session_id", currentVal: current.session_id },
      { key: "peer_session_id", setOp: "set_peer_session_id", deleteOp: "delete_peer_session_id", currentVal: current.peer_session_id },
      { key: "encapsulation", setOp: "set_encapsulation", deleteOp: "delete_encapsulation", currentVal: current.encapsulation },
      { key: "destination_port", setOp: "set_destination_port", deleteOp: "delete_destination_port", currentVal: current.destination_port },
      { key: "source_port", setOp: "set_source_port", deleteOp: "delete_source_port", currentVal: current.source_port },
      { key: "ip_adjust_mss", setOp: "set_ip_adjust_mss", deleteOp: "delete_ip_adjust_mss", currentVal: current.ip_adjust_mss },
      { key: "ip_arp_cache_timeout", setOp: "set_ip_arp_cache_timeout", deleteOp: "delete_ip_arp_cache_timeout", currentVal: current.ip_arp_cache_timeout },
      { key: "ip_source_validation", setOp: "set_ip_source_validation", deleteOp: "delete_ip_source_validation", currentVal: current.ip_source_validation },
      { key: "ipv6_accept_dad", setOp: "set_ipv6_accept_dad", deleteOp: "delete_ipv6_accept_dad", currentVal: current.ipv6_accept_dad },
      { key: "ipv6_adjust_mss", setOp: "set_ipv6_adjust_mss", deleteOp: "delete_ipv6_adjust_mss", currentVal: current.ipv6_adjust_mss },
      { key: "ipv6_base_reachable_time", setOp: "set_ipv6_base_reachable_time", deleteOp: "delete_ipv6_base_reachable_time", currentVal: current.ipv6_base_reachable_time },
      { key: "ipv6_dup_addr_detect_transmits", setOp: "set_ipv6_dup_addr_detect_transmits", deleteOp: "delete_ipv6_dup_addr_detect_transmits", currentVal: current.ipv6_dup_addr_detect_transmits },
      { key: "ipv6_source_validation", setOp: "set_ipv6_source_validation", deleteOp: "delete_ipv6_source_validation", currentVal: current.ipv6_source_validation },
      { key: "ipv6_address_interface_identifier", setOp: "set_ipv6_address_interface_identifier", deleteOp: "delete_ipv6_address_interface_identifier", currentVal: current.ipv6_address_interface_identifier },
      { key: "mirror_ingress", setOp: "set_mirror_ingress", deleteOp: "delete_mirror_ingress", currentVal: current.mirror_ingress },
      { key: "mirror_egress", setOp: "set_mirror_egress", deleteOp: "delete_mirror_egress", currentVal: current.mirror_egress },
    ];

    for (const field of stringFields) {
      if (field.key in updated) {
        const newVal = updated[field.key] as string | null | undefined;
        if (newVal) {
          operations.push({ op: field.setOp, value: newVal });
        } else if (field.currentVal) {
          operations.push({ op: field.deleteOp });
        }
      }
    }

    // Boolean flags
    const booleanFields: { key: keyof typeof updated; currentKey: keyof L2TPv3Interface; setOp: string; deleteOp: string }[] = [
      { key: "disabled", currentKey: "disable", setOp: "set_interface_disable", deleteOp: "delete_interface_disable" },
      { key: "ip_disable_arp_filter", currentKey: "ip_disable_arp_filter", setOp: "set_ip_disable_arp_filter", deleteOp: "delete_ip_disable_arp_filter" },
      { key: "ip_disable_forwarding", currentKey: "ip_disable_forwarding", setOp: "set_ip_disable_forwarding", deleteOp: "delete_ip_disable_forwarding" },
      { key: "ip_enable_arp_accept", currentKey: "ip_enable_arp_accept", setOp: "set_ip_enable_arp_accept", deleteOp: "delete_ip_enable_arp_accept" },
      { key: "ip_enable_arp_announce", currentKey: "ip_enable_arp_announce", setOp: "set_ip_enable_arp_announce", deleteOp: "delete_ip_enable_arp_announce" },
      { key: "ip_enable_arp_ignore", currentKey: "ip_enable_arp_ignore", setOp: "set_ip_enable_arp_ignore", deleteOp: "delete_ip_enable_arp_ignore" },
      { key: "ip_enable_directed_broadcast", currentKey: "ip_enable_directed_broadcast", setOp: "set_ip_enable_directed_broadcast", deleteOp: "delete_ip_enable_directed_broadcast" },
      { key: "ip_enable_proxy_arp", currentKey: "ip_enable_proxy_arp", setOp: "set_ip_enable_proxy_arp", deleteOp: "delete_ip_enable_proxy_arp" },
      { key: "ip_proxy_arp_pvlan", currentKey: "ip_proxy_arp_pvlan", setOp: "set_ip_proxy_arp_pvlan", deleteOp: "delete_ip_proxy_arp_pvlan" },
      { key: "ipv6_disable_forwarding", currentKey: "ipv6_disable_forwarding", setOp: "set_ipv6_disable_forwarding", deleteOp: "delete_ipv6_disable_forwarding" },
      { key: "ipv6_address_autoconf", currentKey: "ipv6_address_autoconf", setOp: "set_ipv6_address_autoconf", deleteOp: "delete_ipv6_address_autoconf" },
      { key: "ipv6_address_no_default_link_local", currentKey: "ipv6_address_no_default_link_local", setOp: "set_ipv6_address_no_default_link_local", deleteOp: "delete_ipv6_address_no_default_link_local" },
    ];

    for (const field of booleanFields) {
      if (updated[field.key] !== undefined) {
        const was = (current[field.currentKey] as boolean | null) ?? false;
        const will = (updated[field.key] as boolean | null) ?? false;
        if (will !== was) {
          operations.push({ op: will ? field.setOp : field.deleteOp });
        }
      }
    }

    // Array: addresses
    if (updated.addresses !== undefined) {
      for (const addr of current.addresses) {
        operations.push({ op: "delete_interface_address", value: addr });
      }
      for (const addr of updated.addresses) {
        operations.push({ op: "set_interface_address", value: addr });
      }
    }

    // Array: ipv6 eui64
    if (updated.ipv6_address_eui64 !== undefined) {
      for (const prefix of current.ipv6_address_eui64) {
        operations.push({ op: "delete_ipv6_address_eui64", value: prefix });
      }
      for (const prefix of updated.ipv6_address_eui64) {
        operations.push({ op: "set_ipv6_address_eui64", value: prefix });
      }
    }

    return this.batchConfigure(name, operations);
  }

  async deleteInterface(name: string): Promise<VyOSResponse> {
    return this.batchConfigure(name, [{ op: "delete_interface" }]);
  }
}

export const l2tpv3Service = new L2TPv3Service();
